Components loading

Daynvheur opened the pull request:
Ability to load and store in URL location the components used.
Limits the number of components/points stored in location to 256, as there is a maximum an url can handle (337, in my Edge tests).

This PR inherits the previous PR's history, but the `c922f2d` "REVERT AutoFFT" commit got main repo's sources back.

Feel free to review and comment any changes.

Daynvheur committed:
const fftUnderSize

Daynvheur committed:
const fftUnderSize

Daynvheur committed:
parameter & complexity max= fftUndersize

Daynvheur committed:
measure the initControls redraw() cost to adapt to low-end hardware

Daynvheur committed:
set controls adapted values

Daynvheur committed:
Lint compliance.

Daynvheur committed:
Lint compliance
I forgot the `const` validations.

Daynvheur committed:
added autofft parameter
Easter-egg: combo with fftsize parameter

Daynvheur committed:
redraw(implexity, intrameter)

Daynvheur committed:
minimal usefull fftSize = 4

Daynvheur committed:
fft clean reinitialization

Daynvheur committed:
added pathReinitialization() method

Daynvheur committed:
\r\n -> \n

Daynvheur committed:
shrinking -> growing fftSize
Starting from 4, max limit fixed at 4096

Daynvheur committed:
Merge branch 'main' into main

Daynvheur committed:
[WIP] Default controls settings reparation

Daynvheur committed:
Cannot use namespace 'FFT' as a type.

Daynvheur committed:
Merge branch 'main' of https://github.com/Daynvheur/epic into main

Daynvheur committed:
moved const lines from global to local

Daynvheur committed:
repaired visiting the page on your phone with no path initially present

Daynvheur committed:
Variable 'fft' implicitly has an 'any' type.

Daynvheur committed:
@typescript-eslint/no-explicit-any
prefer-const

Daynvheur committed:
underscored global complexity and parameter names
Permits redraw without actual complexity/parameter reset.
(+implexity/intrameter disparition)

Daynvheur committed:
[OPTIONAL] arrowification
Method definition HAVE TO happen before use

Daynvheur committed:
[OPTIONAL] - reordered addPoint after input/output sampling.

Daynvheur committed:
@typescript-eslint/semi
@typescript-eslint/space-infix-ops

Daynvheur committed:
re-disabled `redraw` calls from inside initControl

Daynvheur committed:
added resetFft() method

Daynvheur committed:
[LANOITPO] initControls() -> computeRedraw()
Reverted OPTIONAL arrowfication.
computeRedraw() doesn't compute accurately, either too agressively, either too lazily, it's then only set at (first) page's load.
Computing the FFT requires too much FFT recalculation to be done in the draw time of a mouse movement.

Daynvheur committed:
Lint compliance.
See you!

Daynvheur committed:
REVERT AutoFFT

Daynvheur committed:
save-points-button

Daynvheur committed:
setPointsLocaltion

Daynvheur committed:
save_components-button + setComponentsLocation + 'cp' location
Limited up to 256 components/pts in the URL (337 max allowed on Edge)

Daynvheur committed:
compile.bat
Windows CLI batch helper for `parcel serve src/index.html`

Daynvheur committed:
lint compliance

Daynvheur committed:
There was an end of line missing here.

jnm2 commented:
Hey, let me get main building again and I'll check this out

jnm2 committed:
Update all dependencies

jnm2 committed:
Mark entry script as module for Parcel

Parcel treated src/index.ts as a classic browser script because index.html loaded it without type="module". In that mode, imports/exports are invalid and yarn build failed with 'Browser scripts cannot have imports or exports'.

Set the script tag to type="module" so Parcel compiles the TypeScript entry as an ES module and build succeeds.

jnm2 committed:
Update to latest GitHub Actions versions to fix deprecation error

jnm2 committed:
Switch TypeScript to bundler module resolution

TypeScript 6 deprecates the legacy node/node10 moduleResolution mode, which caused yarn validate to fail.

This project is built with Parcel for the browser, so setting moduleResolution to bundler with module set to esnext aligns TypeScript with the actual bundling/runtime model and removes the deprecation failure without suppressing warnings.

jnm2 committed:
Migrate eslint config for v10

jnm2 commented:
The force push was me dropping the commit in the main.yml action to resolve the merge conflict.

jnm2 commented:
"Save points" seems good to me though I'd rather base64url-encode it all. Why have "save components" though? It seems lossy.

Daynvheur commented:
Hello, thanks for the merge.
Yeah, the "save components" is lossy, but it allows for (naive) points sharing.
Base-64 would be better suited, I may add that as a more robust alternative.

Commit:
modernized compile.bat
Plus displayed REM example

Commit:
.vs directory ignorance

Commit:
ignore .cache folder aswell

Commit:
[OPTIONAL] minor code adjustments
Two-liner ifs, early-return if, space between cases
MAJOR CHANGE: .substr -> .substring

Commit:
added raw vs b64 export for points and components
Theyr are still limited to 256 values each, because thousands of values still requires a lots of characters in base64.
Leeway could be added for base64 tho.

Note: Value lists should be more compact (backwards compatibility untouched)

Commit:
ESLint correction.

Daynvheur commented:
I've rebased my main branch onto this one, gathering your updates.

Added modernization to compile.bat (which I think you are not using?), and base64 support (alongside raw).
I must admit that it was a long time I did not tried those, and eventho lossy it's kind of fun to fiddle around raw data.

Daynvheur committed:
Merge branch 'jnm2:main' into main

Commit:
missing double dots on "Save components:"

Commit:
added Managers to tidy the code and [AUTOFFT]adaptFft
Now Fft lowers/raises to sensible values depending on the hardware and cpu load

Commit:
ESLint correction.
Bed time now, whoooo.

Daynvheur commented:
I registered onto Vercel, this looks like a good host.
With byte64 encode: `https://epic-nine-sooty.vercel.app/?range=0&complexity=14&circles=1&encode=btoa;pt;AADiQwAAMEQAwBREAADOQwDACEQAALtDAAAWRAAApEMAgCNEAIC0QwAAFEQAAM9DAAD6QwAA9UMAgNZDAIDjQwCACEQAgL1DAADQQwAAt0MAANlDAICTQwDAFUQAAKRDAIAaRAAAVkMAQCxEAABvQwAAJEQAgLJDAABDRAAAuEMAQEFEAADYQwBAFkQAAM9DAEAWRACAz0MAgOZDAEAvRADAA0QAwBhEAEALRAAAIUQ=`
vs (new) raw : `https://epic-nine-sooty.vercel.app/?range=0&complexity=14&circles=1&pt=|452;704|595;412|547;374|600;328|654;361|592;414|500;490|429;455|546;379|416;366|434;295|599;328|618;214|689;239|656;357|780;368|773;432|601;414|601;415|461;701|527;611|557;644`

Byte64 encoding is more relevant with a large(r) amount of points. Still, it may need a lot of improvements to better shrink the volume of data down: Each point is expected to be in the `short` range (+/-32767 px) while screen panels aren't sized more than 32K (1K is 1920, 4K is 3840, 8K is 7680 and may come in near future, 16K is 15360, 32K is 30720 wide), which may half the byte encoding size.

jnm2 commented:
What about a difference-based encoding, where we encode the x and y diffs from the prev point instead of absolute? Then we could even do a variable-length integer encoding like [BinaryWriter.Write7BitEncodedInt](https://source.dot.net/#System.Private.CoreLib/src/runtime/src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs,2daa1d14ff1877bd) does?

Lmk if you just want to merge too. I don't want to extract more energy than you have atm. I appreciate your interest!

I'd love to break out separate .ts files for separate classes but that can be later.

jnm2 commented:
I approved deployment for this PR. This gives you and me both a hosted preview. I'm curious if it will autodeploy for this PR from now on when you push again.

Daynvheur commented:
I'm interested in the difference-based encoding, the current tests to _code_ points into Int16Array are not yet complete.
And for sure, classes were added for a cleaner separation of concerns, they might be moved into dedicated files.

Feel free to merge when the current state is ready for you, iterations like separation of classes may come as separate  future PRs.
Maybe we should focus on the UI improvement next?

Edit 1: I had a look at Write7BitEncodedInt (and its Write7BitEncodedInt64 sibling), that's clever: the number is read by 7bit chunks (when available), then the remainder (which fits in less than a 7b chunk) is written. That's very inspiring.
Edit 2: Thinking users will usually make small mouse movements, the relative path is tempting, but it requires carefull treatment, as the data chunk only allows for so much points. That means leaving some behind, which is not a problem in absoluteness, but may prove spicy with relative values.
There is no gain in managing the `points` list as relative (those are numbers, they'll occupy the same amount of memory print whatever relatives of absolutes). Same, the raw data may prove difficult to tinker if pathes are not absolutes. But as location encoded value, it's a valid use case, I'll figure something!

Daynvheur committed:
ParameterMangaer -> ParametersManager

jnm2 commented:
Just curious, what is the limitation you're referring to when you say that the data chunk only allows for so many points?

Daynvheur commented:
I'm referring to url length limitations: I don't know how many points will fit into the (compressed) base64 encoding. This might be a lot (more than currently), but it definitely will hit some limit at a point.

Daynvheur commented:
After trying, I don't know why I expected the X/Y positions to be integers, they are indeed in the "short" range, but only for their integer part. I'll try to get something along either a double encoding [integer+'.'+float], or directly converting double->unsigned long if the process can be done in the reverse.

jnm2 commented:
X and Y positions can probably be turned into integers without quantizing (i.e. without precision loss) by taking the DPI into consideration to get physical pixels instead of css pixels?

jnm2 commented:
URLs have no maximum length, and I haven't noticed sites such as mermaid.ai or lab.razor.fyi running into any trouble with putting all the data in the url. Looks like Chrome and Edge support up to 2m characters and Firefox only 1m. But maybe we dont have to care?

Daynvheur committed:
setPointLength -> setPointSegment

Daynvheur committed:
i += n -> i++

Daynvheur committed:
added btoa-na for relative points

Daynvheur committed:
ESLint compliance.
I have to install a local ESLint validator. :)

Daynvheur commented:
URLs def have a maximal length: I had problem copy/pasting a 170ko datablock into Edge.
At worse, there is a practical limit we should aim for to not bloat the Locator more than required. Reminder: the scaling aims at taking consistent index sparsing (if needed) so data degradation should be a problem only on very high dot-precision scenes.
But in the grand scheme of things, yeah we may not care about that and drop this limitation at some point.

I will check X/Y 'real' pixels instead of dpi ones, but not so sure how to start with it (apart from the canvas input area).

Maybe this PR can be merged so the dpi/base64 optimization vs UI can happen on separate PRs?

Daynvheur committed:
more precise adaptFft low/high threshold
But that may need tweaking for other hardware.

Daynvheur committed:
offset -> client

Daynvheur committed:
ESLint recompliance

Daynvheur commented:
`e.clientX`/`e.clientY` look like integer values, they do respect zoomIn/Out aswell, we may use them.

And now, the base64 part.

Daynvheur commented:
I think we can't have variable-length integer encoding. Or not very consistent.

This is what I expected working:
```ts
function encodeBase64(value: number) { // value is expected to be an integer
    if (value < 0)
        value = ~value + 1; // 2-Complement the value if negative

    let binary = '';
    while (value > 127) { // 127 does a good job at being 0x7Fu
        binary += String.fromCharCode(value | ~127); // I do hope ~127 is not calculated at each loop
        value >>= 7;
    }

    return binary + String.fromCharCode(value); // This returns a string of 1-5 chars length, formed by any codes in the range of 0x00u~0x7Fu
}

function decodeBase64(binary: string) {
    let value: number = 0;

    for (let i = 0; i < binary.length; i++)
        value = (value << 7) | binary.charCodeAt(i);
    debugger;
    if (binary.length === 5 && Number(binary.charCodeAt(0)) > 127)
        value = ~value + 1; // 2-Complement if negative

    return value; // This returns the number as-if it was encoded by a string of 0~5 chars length
}
```

With an actual list of points, we may end up with multiple zeroes (like in `64`) which aren't usable as part separation. And any character used as part separation may (and will) correspond to a valid number encoded character. So, basically I think I may use a fixed 5 characters length per point x and y fields, which will not help reduce the overall size.
There has to be another way around this "character separated value set".

I have to step back and think again why we used Float32Array buffers again.

jnm2 commented:
> Maybe this PR can be merged so the dpi/base64 optimization vs UI can happen on separate PRs?

I am fine with that. Let me know and I'll merge without the optimization. Since you worked on it anyway, I'll respond with how I'd think of it.

I would think of the points data as being base64url-encoded (base64url is different than base64) binary. Ignoring its actual representation for now, I would picture just the raw bytes that are being held as points data. The first byte could be a variable-width-encoded int giving us the number of points total, and then we could read that number of pairs of variable-width-encoded ints.

How could we do that? One way could be to use a buffer and fill the bytes with some sort of writer that tracks the write position. It could be based on an array of Uint8Arrays that grows over time, and the Uint8Arrays could be sized so that they could be fed directly to some base64url conversion without having to track partial state between chunks.

That's just one option, maybe it's overkill for the performance needs here. Maybe just pushing bytes to a regular array is fine. Maybe we pull in libraries :D

Daynvheur commented:
Another way to have the variable length working is to dedicate 1 byte in front of the variable-length byte, sort of [Number_of_bytes|1st_byte|..|N-1_byte|Nth_byte], this at best would be a +1/5 overhead, at worse a +1 overhead.
This way, we would know what length to expect. And eventually, it could carry the negative marker to avoid that blind "maxlength+highest part greater than 127".
Useless future-proof side-effect: if the Number go from Int32 to Int64, the Nth part would automagically adjust to represent the greater byte count.
Useless future-future obsolescence: when the Number will go from Int64 to a Int35536 (or anything that is more than a byte count of parts), this prefix will need a refactor. I propose something along a variable-length prefix, with a pre-prefix stating the length of the prefix. (🙂)

Daynvheur commented:
Let's go to main!

That Base64 optimization will happen in another PR.
And I have other WIP things to merge that are in more or less broken state, awaiting for peer review.

jnm2 commented:
(https://github.com/jnm2/epic/pull/13) I decided to drop the idea of LEB128 encoding because I realized that the points aren't integers. I thought maybe we could quantize pixels to integer coordinates to get nice difference-based coding, but this is noticeable distortion on higher zoom levels such as mobile unless something is done to the coordinate system to quantize the pointer location to canvas pixels. 

<img width="771" height="765" alt="Screenshot_20260622_210126_Brave" src="https://github.com/user-attachments/assets/0c508b9e-4fa2-4803-b9a0-596d36107b8a" />

I'm backing away from fancy integer encodings for now. If we really run into pressure to compress, we can explore a bunch of options. For now it seems overengineered.
